projects/hathor-wallet-mobile/security/0001-improved-pin-security-mode.md

```diff
@@ -0,0 +1,66 @@
+
+# Motivation
+[motivation]: #motivation
+
+When using biometrics the pin is saved on the phone system, meaning the storage data is still encrypted with the 6 digit pin, using a longer and more random pin makes the storage data harder to crack.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When enabling the biometrics on the mobile wallet we just set a flag and the pin is always saved on the device keychain, this makes it so enabling and disabling biometrics does not require the user to input the pin.
+This approach makes it easier to set up biometrics and allows the usual pin screen to be used as a fallback.
+
+The new configuration would use the same configuration for biometrics but use a longer and unguessable password (random 32 bytes for instance), it will trust the system prompt completely, so being locked out due to too many attemps will lock the user out of the wallet.
+
+## Access control
+
+The dependency used to manage keychain password is [react-native-keychain](https://www.npmjs.com/package/react-native-keychain) and we use the [setGenericPassword](https://github.com/oblador/react-native-keychain?tab=readme-ov-file#setgenericpasswordusername-password--accesscontrol-accessible-accessgroup-service-securitylevel-) and [getGenericPassword](https://github.com/oblador/react-native-keychain?tab=readme-ov-file#getgenericpassword-authenticationprompt-service-accesscontrol-) to save and retrieve the pin from the keychain.
+Since the new security mode and the current biometrics mode cannot work at the same time we will only use 1 password saved at a time on the keychain (instead of having both pins saved).
+
+The [access control](https://github.com/oblador/react-native-keychain?tab=readme-ov-file#keychainaccess_control-enum) and [accessible](https://github.com/oblador/react-native-keychain?tab=readme-ov-file#keychainaccessible-enum) configurations are used with the `setGenericPassword` to manage how this password can be used and by whom.
+
+This new password should be saved with:
+
+- Access control **`BIOMETRY_ANY_OR_DEVICE_PASSCODE`**: Constraint to access an item with Touch ID for any enrolled fingers or passcode.
+	- This way either biometrics or system pin/pattern will unlock the wallet
+- Accessible **`WHEN_UNLOCKED_THIS_DEVICE_ONLY`**: The data in the keychain item can be accessed only while the device is unlocked by the user. Items with this attribute do not migrate to a new device.
+
+## Settings
+
+The new configuration should not replace the current one, it will be presented to the user as an option called "Enable system authentication" on the settings page, once clicked the enabling process will start on a new screen.
+
+### Enabling process
+
+When trying to enable system authentication the user will be required to confirm that this is what he wants, this screen also has a more detailed explanation of what will be enabled along with a warning that the wallet pin unlock will be disabled.

```

Comment:
Does it make sense to have both ways? When I was reading, I thought we would allow only for the user to have the long password with biometrics, or the 6 digit pin without biometrics.

I think it becomes really confusing having this so customizable.

Reply:
Personally I think that this is a good default for security, but not being able to fallback to the pin may cause issues so i included in the design a way to work as a new feature along with the old biometry mode.

If we decide to only have 1 biometry mode available we could use the feature flag that enables this mode to also disable the old mode, so users will only have 1 biometry mode available for them and gradually we could disable the old biometry mode.

Reply:
I like that you added it to the future possibilities.